net-print/kyocera-1x2x-mfp-driver: New ebuild

firegurafiku opened the pull request:
This ebuild installs prorpietary PPDs and filter program which make CUPS
work with old Kyocera printers of 1x2x MFP family:

 - Kyocera FS-1020MFP
 - Kyocera FS-1025MFP
 - Kyocera FS-1040
 - Kyocera FS-1060DN
 - Kyocera FS-1120MFP
 - Kyocera FS-1125MFP

See also:
* https://bugs.gentoo.org/show_bug.cgi?id=627802
* https://codereview.stackexchange.com/questions/172953

mgorny commented on net-print/kyocera-1x2x-mfp-driver/files/rastertokpsl-fix.sh:
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
```
Are you the author of this script? If not, please indicate the source.

mgorny commented on net-print/kyocera-1x2x-mfp-driver/kyocera-1x2x-mfp-driver-1.1203-r1.ebuild:
```diff
@@ -0,0 +1,131 @@
+SLOT="0"
+
+# TODO: Someone please test the ebuild on ~x86.
+KEYWORDS="-* ~amd64 ~x86"
```
It'd be better if you didn't add `~x86` keyword then, and we can open keywordreq if someone really needs it. However, 32-bit x86 systems are pretty scarce these days, so it's probably not worth the effort.

mgorny commented on net-print/kyocera-1x2x-mfp-driver/kyocera-1x2x-mfp-driver-1.1203-r1.ebuild:
```diff
@@ -0,0 +1,131 @@
+# TODO: Someone please test the ebuild on ~x86.
+KEYWORDS="-* ~amd64 ~x86"
+
+IUSE_LINGUAS=(ar cs de el en es fr he hu it ko pl pt ro ru th tr vi zh_CN zh_TW)
```
`LINGUAS` is deprecated because it violates EAPI 5+ principles. Use `L10N` instead.

mgorny commented on net-print/kyocera-1x2x-mfp-driver/kyocera-1x2x-mfp-driver-1.1203-r1.ebuild:
```diff
@@ -0,0 +1,131 @@
+
+RDEPEND="net-print/cups"
+DEPEND="app-arch/unzip"
+RESTRICT="fetch"
```
I think you want `mirror` here as well.

mgorny commented on net-print/kyocera-1x2x-mfp-driver/kyocera-1x2x-mfp-driver-1.1203-r1.ebuild:
```diff
@@ -0,0 +1,131 @@
+
+S="${WORKDIR}/Linux"
+
+function archSubdirectory {
```
Don't use `function` keyword, and don't do camelcase. It's not Java.

firegurafiku replied:
I know shell is not Java. I'm fine to rewrite the code, but there is some rationale behind these specific design choices:
* camel-case identifiers are extremely unlikely to collide with anything provided by an eclass, shell or external commands.
* keyword `function` lets one easily distinguish between build phases (or other eclass functions to override) and mere helpers.

mgorny commented on net-print/kyocera-1x2x-mfp-driver/kyocera-1x2x-mfp-driver-1.1203-r1.ebuild:
```diff
@@ -0,0 +1,131 @@
+S="${WORKDIR}/Linux"
+
+function archSubdirectory {
+	if use x86 && ! use amd64
```
if ...; then
      ...
    else
      ...
    fi

Please follow the coding style used in other ebuilds.

Besides, this condition doesn't make much sense since arch flags are mutually exclusive. You should also look at `usex` function.

firegurafiku replied:
> Besides, this condition doesn't make much sense since arch flags are mutually exclusive. 

Is it still true in a multiarch installation?

mgorny commented on net-print/kyocera-1x2x-mfp-driver/kyocera-1x2x-mfp-driver-1.1203-r1.ebuild:
```diff
@@ -0,0 +1,131 @@
+function languageArchiveName {
+	# Note the capitalisation inconsistency. Don't "fix" that.
+	case "$1" in
+		ar)    echo arabic      ;;
```
Don't align stuff, and don't space indent.

mgorny commented on net-print/kyocera-1x2x-mfp-driver/kyocera-1x2x-mfp-driver-1.1203-r1.ebuild:
```diff
@@ -0,0 +1,131 @@
+
+pkg_nofetch() {
+	einfo "Please, navigate your browser to the following URL and manually"
+	einfo "download the file named '$A', then put it into '$DISTDIR'."
```
Always use `${...}` for consistency.

mgorny commented on net-print/kyocera-1x2x-mfp-driver/kyocera-1x2x-mfp-driver-1.1203-r1.ebuild:
```diff
@@ -0,0 +1,131 @@
+	einfo "Please, navigate your browser to the following URL and manually"
+	einfo "download the file named '$A', then put it into '$DISTDIR'."
+	einfo "http://www.kyoceradocumentsolutions.eu/index/service/dlc.false.driver.FS1025MFP._.EN.html"
+	einfo ""
```
No need for the empty argument.

mgorny commented on net-print/kyocera-1x2x-mfp-driver/kyocera-1x2x-mfp-driver-1.1203-r1.ebuild:
```diff
@@ -0,0 +1,131 @@
+		local orig="/usr/lib/cups/filter/rastertokpsl"
+		local repl="/usr/libexec/cups/filter/rastertokpsl"
+		use rastertokpsl-fix &&
+			repl="${repl}-fix.sh"
```
`+=`

mgorny commented on net-print/kyocera-1x2x-mfp-driver/kyocera-1x2x-mfp-driver-1.1203-r1.ebuild:
```diff
@@ -0,0 +1,131 @@
+		use rastertokpsl-fix &&
+			repl="${repl}-fix.sh"
+
+		sed -i "s|$orig|$repl|g" "$@"
```
All external commands need `||die`

mgorny commented on net-print/kyocera-1x2x-mfp-driver/kyocera-1x2x-mfp-driver-1.1203-r1.ebuild:
```diff
@@ -0,0 +1,131 @@
+
+		local language="$(languageArchiveName "$lingua")"
+
+		unpack "$g/$language.tar.gz"
```
`unpack` calls really belong in `src_unpack()`.

firegurafiku replied:
Unfortunately, the file provided by Kyocera is a ZIP archive containing a number of tarballs (one for each language). Does unpacking nested tarballs count as 'unpack' or 'prepare'? We don't potentially need to unpack them all. Is it okay for `src_unpack` to depend on use flags?

mgorny commented on net-print/kyocera-1x2x-mfp-driver/kyocera-1x2x-mfp-driver-1.1203-r1.ebuild:
```diff
@@ -0,0 +1,131 @@
+
+		# The upstream contains a number of files named 'rastertokpsl',
+		# each having the same content.
+		mv --no-clobber "$language/rastertokpsl" "$S/rastertokpsl"
```
`--no-clobber` is not portable.

mgorny commented on net-print/kyocera-1x2x-mfp-driver/kyocera-1x2x-mfp-driver-1.1203-r1.ebuild:
```diff
@@ -0,0 +1,131 @@
+}
+
+src_install() {
+
```
Remove the stray line.

mgorny commented on net-print/kyocera-1x2x-mfp-driver/kyocera-1x2x-mfp-driver-1.1203-r1.ebuild:
```diff
@@ -0,0 +1,131 @@
+		local lingua="$1"; shift
+		for file in "$@"; do
+			local dn="$(dirname "$file")"
+			local bn="$(basename "$file" .ppd)"
```
Don't use `dirname`, `basename`; use bash variable substitutions (`%/*`, `##*/`) instead.

mgorny commented on net-print/kyocera-1x2x-mfp-driver/kyocera-1x2x-mfp-driver-1.1203-r1.ebuild:
```diff
@@ -0,0 +1,131 @@
+		for file in "$@"; do
+			local dn="$(dirname "$file")"
+			local bn="$(basename "$file" .ppd)"
+			newins "$dn/$bn.ppd" "${bn}_$lingua.ppd"
```
Isn't `"$dn/$bn.ppd"` basically `"${file}"`?

mgorny commented on net-print/kyocera-1x2x-mfp-driver/kyocera-1x2x-mfp-driver-1.1203-r1.ebuild:
```diff
@@ -0,0 +1,131 @@
+
+	function installWithLinguaSuffix {
+		local lingua="$1"; shift
+		for file in "$@"; do
```
`in "$@"` is unnecessary, `for file; do` does the same.

mgorny commented on net-print/kyocera-1x2x-mfp-driver/kyocera-1x2x-mfp-driver-1.1203-r1.ebuild:
```diff
@@ -0,0 +1,131 @@
+
+	dodoc "Readme.htm"
+
+	einfo "This packages installs Kyocera's 'Readme.htm' file (see package"
```
I think you want `elog` here, since the user probably wants to see this after the install finished. `einfo` is meant for status messages.

mgorny commented on net-print/kyocera-1x2x-mfp-driver/metadata.xml:
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE pkgmetadata SYSTEM "http://www.gentoo.org/dtd/metadata.dtd">
+<pkgmetadata>
```
List yourself as the maintainer, and proxy-maint project to commit for you.

firegurafiku commented:
@mgorny 

Many thanks for your code review, I'll try to fix all the issues ASAP.
By the way, what is the desired behavior when user has no `L10N` languages set, or their languages are unsupported by the package?

mgorny commented:
I suppose REQUIRED_USE enforcing at least one being selected would make most sense. Make sure to put english first so GLEP73 algo chooses it if no language is selected.

firegurafiku commented:
@mgorny 

Hi! I rewrote the script in a bit different fashion; that (hopefully) fixed all the issues you had noted.
Can you please give it another look?

By the way, is there a way to suppress the following message? This is a proprietary binary, it's supposed to be already stripped.

    * QA Notice: Pre-stripped files found:
    * /usr/libexec/cups/filter/rastertokpsl

mgorny commented:
Look at `man 5 ebuild`, QA Control Variables section (`QA_PREBUILT` in particular).

mgorny commented on net-print/kyocera-1x2x-mfp-driver/files/rastertokpsl-fix.sh:
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+# This file is a slightly modified version of 'rastertokpsl_wrapper' script
+# by Maxim Norin. See https://github.com/mnorin/kyocera-ppd-installer for
```
Actually, the upstream repo indicates that the wrapper has a different author. But I guess it's acceptable as long as there's URI. I've filed mnorin/kyocera-ppd-installer#2 to get clear licenses statement from upstream.

mnorin replied:
Please check https://github.com/mnorin/kyocera-ppd-installer for licenses information and let me know if everything is clear enough

firegurafiku replied:
@mnorin 

Thank you, the license is clear and seems to compatible with portage. However, if the code originally had a different author, we should mention him as well (an ask if he is fine with GPLv2). This makes things complicated :(.


@mnorin

mnorin replied:
The only information I have about wrapper author is the link to Yandex.Market comment https://market.yandex.ru/product/8523545/forum?hid=138608&comment=child-7-1477811

mgorny commented on net-print/kyocera-1x2x-mfp-driver/kyocera-1x2x-mfp-driver-1.1203-r1.ebuild:
```diff
@@ -0,0 +1,133 @@
+}
+
+src_unpack() {
+	unpack $A
```
`${A}`, i.e. use {} always for consistency.

mgorny commented on net-print/kyocera-1x2x-mfp-driver/kyocera-1x2x-mfp-driver-1.1203-r1.ebuild:
```diff
@@ -0,0 +1,133 @@
+		sed -i "s|${orig}|${repl}|g" "$@" || die
+	}
+
+	patch_filter_path **/Kyocera*.ppd
```
Are you relying on globstar here or is the double `**` accidental?

firegurafiku replied:
Yes, it's an intentional globstar expansion. Should I avoid it (I can)?

firegurafiku commented:
@mgorny

I added `QA_PREBUILT`, fixed forgotten `${...}` and removed globstar expansion (is it evil?).
By the way, is it fine to call `find -print -quit` from an ebuild?

mgorny commented:
`globstar` is not a problem in itself but you need to explicitly enable it before using, and restore previous state after use. `find` is fine for use, and is guaranteed to be compatible with GNU find.

mgorny commented on net-print/kyocera-1x2x-mfp-driver/kyocera-1x2x-mfp-driver-1.1203-r1.ebuild:
```diff
@@ -0,0 +1,134 @@
+
+	local l10n
+	for l10n in "${IUSE_L10N[@]}"; do
+		if use "l10n_${l10n}"; then
```
Actually, in EAPI 6 you can just do:

    for l10n in ${L10N}; do ...

since `${L10N}` is set to contain all enabled flags from the USE_EXPAND.

firegurafiku replied:
Great hint, thanks!

mgorny commented on net-print/kyocera-1x2x-mfp-driver/kyocera-1x2x-mfp-driver-1.1203-r1.ebuild:
```diff
@@ -0,0 +1,134 @@
+		sed -i "s|${orig}|${repl}|g" "$@" || die
+	}
+
+	patch_filter_path ./*/Kyocera*.ppd
```
Actually, why do you need a sub-function for this? After all, you just pass all paths to sed, so you could just inline it.

firegurafiku replied:
From my non-shell programming background, having a separate function for a well-defined task feels like a "right" thing even if I have to call it only once. If only Bash had the notion of closures and nested functions 😢. (The fact that this "nested" functions are actually global made me name them in camel case originally, so nothing can interfere with them by accident.)

I don't insist on having this as a function here, but I believe it's making the code more readable. The same also applies for `install_with_l10n_suffix`.

mgorny replied:
Well, lack of local functions is exactly why I don't like it here. It's causing global env pollution for no real advantage.

firegurafiku replied:
BTW, it's `kyocera-mita-ppds-8.4.ebuild` where I borrowed the "use nested function if you want"-approach. I thought it's a kind of well-established practice in the portage tree.

firegurafiku replied:
I'm fine with inlining `patch_filter_path`, but getting rid of `install_with_l10n_suffix` will make the code look much worse. Can we just prefix the name with underscore (like in Python), or `aux_`, or something else to denote it's for internal use? 

Or file a patch to the Bash mailing list to get actual support for nested functions ^__^.

mgorny replied:
The latter function is fine, it at least has a valid case for calling multiple times. To make it feel more local, you can `unset -f` it after use.

mgorny commented on net-print/kyocera-1x2x-mfp-driver/kyocera-1x2x-mfp-driver-1.1203-r1.ebuild:
```diff
@@ -0,0 +1,134 @@
+
+	# Each unpacked tarball contains an exact copy of 'rastertokpsl',
+	# just take any of them.
+	local rasterfile="$(find -name rastertokpsl -print -quit || die)"
```
You shouldn't die inside a subshell. Better `|| die` outside it (the exit status is preserved post var assignment).

Besides, it's not catching the error when it can't find one. Account for the possibility of null output.

firegurafiku commented:
@mgorny 

Done with inlining. I believe, the patch is almost ready now.
By the way, we should mention @mnorin in the commit message somehow. Probably, a 'Signed-off-by' line, or something similar?

mgorny commented:
No, signoff is for DCO.

Oh, right, you should probably mention GPL-2(+?) in the licenses now.

firegurafiku commented:
Done with adding 'GPL-2'.
BTW, my local `repoman` is complaining about `quoted "${A}" on line: 62`. (I forgot, but that was the reason why I unquoted it.)

mgorny commented on net-print/kyocera-1x2x-mfp-driver/kyocera-1x2x-mfp-driver-1.1203-r1.ebuild:
```diff
@@ -0,0 +1,129 @@
+HOMEPAGE="http://www.kyoceradocumentsolutions.eu"
+SRC_URI="LinuxDrv_${PV}_FS-1x2xMFP.zip"
+
+LICENSE="kyocera-mita-ppds GPL-2"
```
Sorting!

mgorny commented on net-print/kyocera-1x2x-mfp-driver/kyocera-1x2x-mfp-driver-1.1203-r1.ebuild:
```diff
@@ -0,0 +1,129 @@
+}
+
+src_unpack() {
+	unpack "${A}"
```
Yeah, `A` is a space separated list of multiple files; that's why we don't quote it. Yes, it sucks by design but I can't think of anything worth doing about it atm.

mgorny commented on net-print/kyocera-1x2x-mfp-driver/kyocera-1x2x-mfp-driver-1.1203-r1.ebuild:
```diff
@@ -0,0 +1,129 @@
+pkg_nofetch() {
+	einfo "Please, navigate your browser to the following URL and manually"
+	einfo "download the file named '${A}', then put it into '${DISTDIR}'."
+	einfo "http://www.kyoceradocumentsolutions.eu/index/service/dlc.false.driver.FS1025MFP._.EN.html"
```
```
 * Please, navigate your browser to the following URL and manually
 * download the file named 'LinuxDrv_1.1203_FS-1x2xMFP.zip', then put it into '/tmp/tmprn9z_egz/portage/net-print/kyocera-1x2x-mfp-driver-1.1203-r1/distdir'.
```

Not very helpful directory, i.e. you are using `DISTDIR` the way it's not supposed to be used.

mgorny replied:
Also, https here. Especially that the site redirects.

mgorny commented:
Ok, installs good. I'm not going to bother you with another round of review. Just please take my comments into consideration for your future work.

firegurafiku commented:
@mgorny

Thank you for your help, you did a great reviewing job. I hope, my future ebuilds wouldn't require that much attention from you or some other maintainer.
